File: code/generate-sharp/utilities/operation-graph/OperationGraph.cs
// <copyright file="OperationGraph.cs" company="Soup">
// Copyright (c) Soup. All rights reserved.
// </copyright>

using System.Collections.Generic;
using Path = Opal.Path;

namespace Soup.Build.Utilities;

/// <summary>
/// The cached operation graph that is used to track input/output mappings for previous build
/// executions to support incremental builds
/// </summary>
public class OperationGraph
{
	private readonly Dictionary<OperationId, OperationInfo> operations;
	private readonly Dictionary<CommandInfo, OperationId> operationLookup;

	private readonly Dictionary<OperationProxyId, OperationProxyInfo> operationProxies;
	private readonly Dictionary<CommandInfo, OperationProxyId> operationProxyLookup;

	/// <summary>
	/// Initializes a new instance of the <see cref="OperationGraph"/> class.
	/// </summary>
	public OperationGraph()
	{
		this.ReferencedFiles = [];
		this.RootOperationIds = [];
		this.operations = [];
		this.operationLookup = [];
		this.operationProxies = [];
		this.operationProxyLookup = [];
	}

	/// <summary>
	/// Initializes a new instance of the <see cref="OperationGraph"/> class.
	/// </summary>
	public OperationGraph(
		IList<(FileId FileId, Path Path)> referencedFiles,
		IList<OperationId> rootOperations,
		IList<OperationInfo> operations,
		IList<OperationProxyInfo> operationProxies)
	{
		this.ReferencedFiles = referencedFiles;
		this.RootOperationIds = rootOperations;
		this.operations = [];
		this.operationLookup = [];
		this.operationProxies = [];
		this.operationProxyLookup = [];

		// Store the incoming vector of operations as a lookup for fast checks
		foreach (var info in operations)
		{
			AddOperation(info);
		}

		foreach (var info in operationProxies)
		{
			AddOperationProxy(info);
		}
	}

	/// <summary>
	/// Get the set of referenced file ids that map to their paths
	/// </summary>
	public IList<(FileId FileId, Path Path)> ReferencedFiles { get; init; }

	/// <summary>
	/// Get the list of root operation ids
	/// </summary>
	public IList<OperationId> RootOperationIds { get; init; }

	/// <summary>
	/// Get Operations
	/// </summary>
	public IReadOnlyDictionary<OperationId, OperationInfo> Operations => this.operations;

	/// <summary>
	/// Find an operation info
	/// </summary>
	public bool HasCommand(CommandInfo command)
	{
		return this.operationLookup.ContainsKey(command);
	}

	/// <summary>
	/// Find an operation info
	/// </summary>
	public bool TryFindOperationInfo(
		CommandInfo command,
		out OperationInfo operation)
	{
		if (this.operationLookup.TryGetValue(command, out var operationId))
		{
			operation = GetOperationInfo(operationId);
			return true;
		}
		else
		{
			operation = new OperationInfo();
			return false;
		}
	}

	/// <summary>
	/// Get an operation info
	/// </summary>
	public OperationInfo GetOperationInfo(OperationId operationId)
	{
		return this.operations[operationId];
	}

	/// <summary>
	/// Add an operation info
	/// </summary>
	public void AddOperation(OperationInfo info)
	{
		this.operationLookup.Add(info.Command, info.Id);
		this.operations.Add(info.Id, info);
	}

	/// <summary>
	/// Add an operation proxy info
	/// </summary>
	public void AddOperationProxy(OperationProxyInfo info)
	{
		this.operationProxyLookup.Add(info.Command, info.Id);
		this.operationProxies.Add(info.Id, info);
	}
}

File: code/client/core/source/operation-graph/OperationGraph.cpp
﻿// <copyright file="OperationGraph.cpp" company="Soup">
// Copyright (c) Soup. All rights reserved.
// </copyright>

module;

#include <map>
#include <stdexcept>
#include <unordered_map>
#include <vector>

export module Soup.Core:OperationGraph;

import Opal;
import :CommandInfo;
import :OperationInfo;

using namespace Opal;

namespace Soup::Core
{
	/// <summary>
	/// The operation graph that represents the set of operations that need to be evaluated to perform the build
	/// </summary>
	export class OperationGraph
	{
	private:
		std::vector<OperationId> _rootOperations;
		std::map<OperationId, OperationInfo> _operations;
		std::unordered_map<CommandInfo, OperationId> _operationLookup;

	public:
		/// <summary>
		/// Initializes a new instance of the <see cref="OperationGraph"/> class.
		/// </summary>
		OperationGraph() :
			_rootOperations(),
			_operations(),
			_operationLookup()
		{
		}

		/// <summary>
		/// Initializes a new instance of the <see cref="OperationGraph"/> class.
		/// </summary>
		OperationGraph(
			std::vector<OperationId> rootOperations,
			std::vector<OperationInfo> operations) :
			_rootOperations(std::move(rootOperations)),
			_operations(),
			_operationLookup()
		{
			// Store the incoming vector of operations as a lookup for fast checks
			for (auto& info : operations)
			{
				AddOperation(std::move(info));
			}
		}

		/// <summary>
		/// Get the list of root operation ids
		/// </summary>
		const std::vector<OperationId>& GetRootOperationIds() const
		{
			return _rootOperations;
		}

		/// <summary>
		/// Set the list of root operation ids
		/// </summary>
		void SetRootOperationIds(std::vector<OperationId> value)
		{
			_rootOperations = std::move(value);
		}

		/// <summary>
		/// Get Operations
		/// </summary>
		const std::map<OperationId, OperationInfo>& GetOperations() const
		{
			return _operations;
		}

		/// <summary>
		/// Get Operations
		/// </summary>
		std::map<OperationId, OperationInfo>& GetOperations()
		{
			return _operations;
		}

		/// <summary>
		/// Find an operation info
		/// </summary>
		bool HasOperationCommand(const CommandInfo& command) const
		{
			return _operationLookup.contains(command);
		}

		/// <summary>
		/// Find an operation info
		/// </summary>
		bool TryFindOperation(
			const CommandInfo& command,
			OperationId& operationId) const
		{
			auto findResult = _operationLookup.find(command);
			if (findResult != _operationLookup.end())
			{
				operationId = findResult->second;
				return true;
			}
			else
			{
				return false;
			}
		}

		/// <summary>
		/// Get an operation info
		/// </summary>
		OperationInfo& GetOperationInfo(OperationId operationId)
		{
			auto findResult = _operations.find(operationId);
			if (findResult != _operations.end())
			{
				return findResult->second;
			}
			else
			{
				throw std::runtime_error("The provided operation id does not exist");
			}
		}
		const OperationInfo& GetOperationInfo(OperationId operationId) const
		{
			auto findResult = _operations.find(operationId);
			if (findResult != _operations.end())
			{
				return findResult->second;
			}
			else
			{
				throw std::runtime_error("The provided operation id does not exist");
			}
		}

		/// <summary>
		/// Add an operation info
		/// </summary>
		OperationInfo& AddOperation(OperationInfo info)
		{
			auto insertLookupResult = _operationLookup.emplace(info.Command, info.Id);
			if (!insertLookupResult.second)
				throw std::runtime_error("The provided command already exists in the graph");

			auto [insertIterator, wasInserted] = _operations.emplace(info.Id, std::move(info));
			if (!wasInserted)
				throw std::runtime_error("The provided operation id already exists in the graph");

			return insertIterator->second;
		}

		/// <summary>
		/// Equality operator
		/// </summary>
		bool operator ==(const OperationGraph& rhs) const
		{
			return _rootOperations == rhs._rootOperations &&
				_operations == rhs._operations;
		}

		/// <summary>
		/// Inequality operator
		/// </summary>
		bool operator !=(const OperationGraph& rhs) const
		{
			return !(*this == rhs);
		}
	};
}